### PR TITLE
fix frontend to handle the case user owns a domain

### PR DESCRIPTION
**Pull request by 0xvgd**
https://github.com/polymorpher/otc-country/issues/10
https://github.com/polymorpher/otc-country/issues/11

image

How to test:
- Contracts on Sepolia
  - OTC: 0x012BDE70f155A2E2d0C2AF257a4d9eCdBA20653c
  - OTC Factory: 0x8278Dc174962D7ea2dCbbFA4FA3eE474cabB9443
  - Mock domain contract: 0xD81927f7e00A2Ed1eE1815e6D63594d0b52f9f57
  - TokenA: 0x9288Ad4702cF47c44dC714312498B35cd13a2C9a (added to OTC)
  - TokenB: 0x7D7E515fF4e4B82c5EcbD2583Cb7314f8529Bd59 (added to OTC)
  - You can call `mint()` function of TokenA and TokenB to test https://sepolia.etherscan.io/address/0x9288Ad4702cF47c44dC714312498B35cd13a2C9a#writeContract#F5
- Disable frontend ChainDetector to test on Sepolia
- Add these Sepolia ERC20 tokens in https://github.com/polymorpher/otc-country/blob/feat/offer-history/app/src/helpers/assets.ts#L92
  - Token A: 0x9288Ad4702cF47c44dC714312498B35cd13a2C9a
  - Token B: 0x7D7E515fF4e4B82c5EcbD2583Cb7314f8529Bd59
  - These tokens are already added to OTC
- Update frontend .env
  - VITE_OTC_ADDRESS
  - VITE_RPC (alchemy or infura url)
  - VITE_SERVER
- Run server
  - Update server .env
    - OTC_CONTRACT
    - RPC_URL (alchemy or infura url)
  - npm run build
  - node dist/server/listen.js
  - node dist/server/index.js (run on 3000 by default)

### CODE REVIEW

File: contract/contracts/OTC.sol [hunk 1]
```diff
@@ -239,20 +238,19 @@ contract OTC is AccessControl, Pausable {
             revert OTCError(ErrorType.CommissionRateBeyondLimit);
         }
 
-        if (domainContract.ownerOf(domainName_) == domainOwner_) {
-            revert OTCError(ErrorType.AlreadyOwner);
-        }
-
-        uint256 price = domainContract.getPrice(domainName_);
-        bytes32 commitment = domainContract.makeCommitment(domainName_, domainOwner_, secret_);
+        if (domainContract.ownerOf(domainName_) != domainOwner_) {
```
Comment (polymorpher) [inline contract/contracts/OTC.sol]: May revert, need `try` block

File: server/index.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+app.get("/", async (request: Request, response: Response) => {
+  const { asset, age } = request.query
+  const page = Number(request.query.page ?? 0)
+  const pageSize = Number(request.query.pageSize ?? 10)
+
+  const where = ["TRUE"], args = []
+
+  if (asset !== undefined) {
+    where.push("(src_asset = $ OR dest_asset = $)")
+    args.push(asset)
```
Comment (polymorpher) [inline server/index.ts:59]: Sanitize inputs - check and make sure they belong to a list (or at least match address regex)

File: server/listen.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+        const [srcPrice, destPrice, time] = await Promise.all([
+          getPrice(String(srcAsset)),
+          getPrice(String(destAsset)),
+          getTimestamp(Number(log.blockNumber))
+        ])
+
+        await pool.query(`
+          INSERT INTO
+            logs(event_name, time, src_asset, dest_asset, offer_address, domain_owner, close_amount, total_deposits, src_price, dest_price)
```
Comment (polymorpher) [inline server/listen.ts]: Not guarding against duplicates. Should create deterministic id based on data (transaction hash, event position in transaction, ...)

File: server/listen.ts [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        if (log.eventName === 'OfferCreated') {
+          const [srcPrice, destPrice, time] = await Promise.all([
+            getPrice(log.args.srcAsset!),
+            getPrice(log.args.destAsset!),
+            getTimestamp(Number(log.blockNumber))
+          ])
+
+          await pool.query(`
+            INSERT INTO
+              logs(event_name, time, src_asset, dest_asset, offer_address, domain_owner, close_amount, total_deposits, src_price, dest_price)
```
Comment (polymorpher) [inline server/listen.ts]: Same here

File: server/listen.ts [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  return timestampCache[blockNumber]
+}
+
+const listen = async () => {
+  let lastBlockNumber = 0
+
+  if (fs.existsSync(SAVE_FILE_PATH)) {
+    lastBlockNumber = Number(fs.readFileSync(SAVE_FILE_PATH))
```
Comment (polymorpher) [inline server/listen.ts:34]: Server can be stateless - use db and avoid fs if possible

**Review by polymorpher — CHANGES_REQUESTED** [review]
Will continue tomorrow and try frontend

File: contract/contracts/OTC.sol (hunk 1 above)
Comment (0xvgd) [reply contract/contracts/OTC.sol]: What do you mean? When revert?

File: server/listen.ts (hunk 5 above)
Comment (0xvgd) [reply server/listen.ts:34]: Can listen run on stateless server? as it runs from setInterval

File: server/index.ts (hunk 2 above)
Comment (0xvgd) [reply server/index.ts:59]: Done

File: server/listen.ts (hunk 3 above)
Comment (0xvgd) [reply server/listen.ts]: When does the duplication happen?
Duplication will be prevented by `lastBlockNumberBeingProcessed`
https://github.com/polymorpher/otc-country/blob/feat/offer-history/server/listen.ts#L44-L59

Comment (0xvgd) [reply server/listen.ts]: How about adding unique constraint to table?
```
  UNIQUE (event_name, offer_address)
```

Comment (polymorpher) [reply server/listen.ts]: Okay, make it fail-safe - should simply log (to file or console) constraint violation and continue with next record

File: contract/contracts/OTC.sol (hunk 1 above)
Comment (polymorpher) [reply contract/contracts/OTC.sol]: https://github.com/polymorpher/dot-country/blob/14dc510dd9a354826b23f21aced3eb26ee5bba93/contracts/contracts/DC.sol#L194

https://github.com/polymorpher/ens-deployer/blob/e31d1d479a27f18941ee9f5ced1ecfdc168b99f8/contract/contracts/TLDBaseRegistrarImplementation.sol#L115

File: server/listen.ts (hunk 5 above)
Comment (polymorpher) [reply server/listen.ts:34]: Maybe good for now, some of those servers disable fs access. Can fix later when there is a concrete need. Will use a full VM for now

File: server/listen.ts (hunk 3 above)
Comment (0xvgd) [reply server/listen.ts]: Done

File: contract/contracts/OTC.sol (hunk 1 above)
Comment (0xvgd) [reply contract/contracts/OTC.sol]: Updated

**Comment by polymorpher** [comment]
Not letting me use an existing domain

image

**Comment by polymorpher** [comment]
Can you add a script (SQL is fine) to populate some fake data in the database so I can see how offer history looks like and try out all interactions?

**Comment by polymorpher** [comment]
Let's put offer history in a separate page. Add a small section above "Create Offer" that shows most recent 3 offers, in a single row, 3 columns, each in a box (time, amount, domain, from what asset to what asset, exchange rate). Small text is fine within each box

image

**Comment by polymorpher** [comment]
Merging this PR. Let's make the changes in a separate PR

**Comment by 0xvgd** [comment]
> Not letting me use an existing domain

Was testing with Sepolia, or harmony?
If Sepolia, it has mock code like https://github.com/polymorpher/otc-country/blob/main/contract/contracts/mocks/DomainContract.sol#L17-L19

**Comment by polymorpher** [comment]
> > Not letting me use an existing domain
> 
> 
> 
> Was testing with Sepolia, or harmony?
> 
> If Sepolia, it has mock code like https://github.com/polymorpher/otc-country/blob/main/contract/contracts/mocks/DomainContract.sol#L17-L19

On Harmony - production contract

**Comment by 0xvgd** [comment]
> Let's put offer history in a separate page.

How to navigate to the separate offer history? tab?

**Comment by 0xvgd** [comment]
> Not letting me use an existing domain "polymorpher"

I tested, but the message seems correct because:
- domainContract.available("polymorpher") is false
- https://1ns-registrar-relayer.hiddenstate.xyz/check-domain returns false for "polymorpher"

**Comment by polymorpher** [comment]
If it is unavailable and is owned by the user, that means obviously the user intends to use the domain for the offer, so it should let the user proceed with creating the offer

**Comment by polymorpher** [comment]
> > Let's put offer history in a separate page.
> 
> How to navigate to the separate offer history? tab?

The section heading links (opening new tab) to that page

> "Add a small section above "Create Offer" that shows most recent 3 offers, in a single row, 3 columns, each in a box (time, amount, domain, from what asset to what asset, exchange rate). Small text is fine within each box"

**Comment by 0xvgd** [comment]
In what page does it show offer history filter?

**Comment by polymorpher** [comment]
Landing page and `/offers`. New offer creation at `/new`

**Comment by 0xvgd** [comment]
Sorry, can't understand.
Currently, we create a new offer in landing page.
Now we will have 3 columned recent offer in landing page. Show it along with offer filter?

**Comment by polymorpher** [comment]
When users visit the site, they expect to see what offers are available, adjust filters and look at how things change over time, and check some of them individually. It doesn't make sense to show them offer creation page by default. So do not show offer creation here. Instead, only provide a link or button which leads to offer creation page

At offer creation page, it doesn't make sense to show all recent offers or include all features for browsing offers, such as the filter, because they come here to create offers, not to browse offers. So it makes sense to only show three recent offers in small boxes, laid in a single row of three columns. They provide useful information for users to understand market price, when they create their own offer. Provide only a link to navigate to full offer browsing page

**Pull request merged**